File: web/src/__tests__/NativeErrorCode.test.ts
import {
  NativeErrorCode,
  nativeErrorCodeFromInt,
  nativeErrorToEdgeVedaError,
  throwIfNativeError,
} from '../NativeErrorCode';
import { EdgeVedaError } from '../types';

describe('NativeErrorCode', () => {
  describe('enum values', () => {
    it('OK === 0', () => {
      expect(NativeErrorCode.OK).toBe(0);
    });

    it('INVALID_PARAMETER === -1', () => {
      expect(NativeErrorCode.INVALID_PARAMETER).toBe(-1);
    });

    it('OUT_OF_MEMORY === -2', () => {
      expect(NativeErrorCode.OUT_OF_MEMORY).toBe(-2);
    });

    it('MODEL_LOAD_FAILED === -3', () => {
      expect(NativeErrorCode.MODEL_LOAD_FAILED).toBe(-3);
    });

    it('BACKEND_INIT_FAILED === -4', () => {
      expect(NativeErrorCode.BACKEND_INIT_FAILED).toBe(-4);
    });

    it('INFERENCE_FAILED === -5', () => {
      expect(NativeErrorCode.INFERENCE_FAILED).toBe(-5);
    });

    it('CONTEXT_INVALID === -6', () => {
      expect(NativeErrorCode.CONTEXT_INVALID).toBe(-6);
    });

    it('STREAM_ENDED === -7', () => {
      expect(NativeErrorCode.STREAM_ENDED).toBe(-7);
    });

    it('NOT_IMPLEMENTED === -8', () => {
      expect(NativeErrorCode.NOT_IMPLEMENTED).toBe(-8);
    });

    it('MEMORY_LIMIT_EXCEEDED === -9', () => {
      expect(NativeErrorCode.MEMORY_LIMIT_EXCEEDED).toBe(-9);
    });

    it('UNSUPPORTED_BACKEND === -10', () => {
      expect(NativeErrorCode.UNSUPPORTED_BACKEND).toBe(-10);
    });

    it('UNKNOWN === -999', () => {
      expect(NativeErrorCode.UNKNOWN).toBe(-999);
    });
  });

  describe('nativeErrorCodeFromInt', () => {
    it.each([
      [0, NativeErrorCode.OK],
      [-1, NativeErrorCode.INVALID_PARAMETER],
      [-2, NativeErrorCode.OUT_OF_MEMORY],
      [-3, NativeErrorCode.MODEL_LOAD_FAILED],
      [-4, NativeErrorCode.BACKEND_INIT_FAILED],
      [-5, NativeErrorCode.INFERENCE_FAILED],
      [-6, NativeErrorCode.CONTEXT_INVALID],
      [-7, NativeErrorCode.STREAM_ENDED],
      [-8, NativeErrorCode.NOT_IMPLEMENTED],
      [-9, NativeErrorCode.MEMORY_LIMIT_EXCEEDED],
      [-10, NativeErrorCode.UNSUPPORTED_BACKEND],
      [-999, NativeErrorCode.UNKNOWN],
    ])('maps %i → NativeErrorCode %i', (code, expected) => {
      expect(nativeErrorCodeFromInt(code)).toBe(expected);
    });

    it('maps unmapped integer to UNKNOWN', () => {
      expect(nativeErrorCodeFromInt(-42)).toBe(NativeErrorCode.UNKNOWN);
    });

    it('maps positive unmapped integer to UNKNOWN', () => {
      expect(nativeErrorCodeFromInt(99)).toBe(NativeErrorCode.UNKNOWN);
    });
  });

  describe('nativeErrorToEdgeVedaError', () => {
    it('returns null for OK', () => {
      expect(nativeErrorToEdgeVedaError(NativeErrorCode.OK)).toBeNull();
    });

    it('returns null for STREAM_ENDED (sentinel, not an error)', () => {
      expect(nativeErrorToEdgeVedaError(NativeErrorCode.STREAM_ENDED)).toBeNull();
    });

    it('returns EdgeVedaError for INVALID_PARAMETER', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.INVALID_PARAMETER);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for OUT_OF_MEMORY', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.OUT_OF_MEMORY);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for MODEL_LOAD_FAILED', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.MODEL_LOAD_FAILED);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for BACKEND_INIT_FAILED', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.BACKEND_INIT_FAILED);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for INFERENCE_FAILED', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.INFERENCE_FAILED);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for CONTEXT_INVALID', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.CONTEXT_INVALID);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for NOT_IMPLEMENTED', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.NOT_IMPLEMENTED);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for MEMORY_LIMIT_EXCEEDED', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.MEMORY_LIMIT_EXCEEDED);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for UNSUPPORTED_BACKEND', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.UNSUPPORTED_BACKEND);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('returns EdgeVedaError for UNKNOWN', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.UNKNOWN);
      expect(err).toBeInstanceOf(EdgeVedaError);
    });

    it('includes context string in error message when provided', () => {
      const err = nativeErrorToEdgeVedaError(
        NativeErrorCode.INFERENCE_FAILED,
        'token decode'
      );
      expect(err).not.toBeNull();
      expect(err!.message).toContain('token decode');
    });

    it('error message excludes context suffix when context is omitted', () => {
      const err = nativeErrorToEdgeVedaError(NativeErrorCode.INFERENCE_FAILED);
      expect(err).not.toBeNull();
      // Should not have the ": " separator without context
      expect(err!.message).not.toMatch(/:\s*$/);
    });
  });

  describe('throwIfNativeError', () => {
    it('does not throw for OK', () => {
      expect(() => throwIfNativeError(NativeErrorCode.OK)).not.toThrow();
    });

    it('does not throw for STREAM_ENDED', () => {
      expect(() => throwIfNativeError(NativeErrorCode.STREAM_ENDED)).not.toThrow();
    });

    it('throws EdgeVedaError for INFERENCE_FAILED', () => {
      expect(() => throwIfNativeError(NativeErrorCode.INFERENCE_FAILED)).toThrow(
        EdgeVedaError
      );
    });

    it('throws EdgeVedaError for MODEL_LOAD_FAILED', () => {
      expect(() => throwIfNativeError(NativeErrorCode.MODEL_LOAD_FAILED)).toThrow(
        EdgeVedaError
      );
    });

    it('accepts a raw integer and throws for error codes', () => {
      expect(() => throwIfNativeError(-5)).toThrow(EdgeVedaError);
    });

    it('accepts raw integer 0 (OK) without throwing', () => {
      expect(() => throwIfNativeError(0)).not.toThrow();
    });

    it('accepts raw integer -7 (STREAM_ENDED) without throwing', () => {
      expect(() => throwIfNativeError(-7)).not.toThrow();
    });
  });
});

File: kotlin/src/test/kotlin/com/edgeveda/sdk/NativeErrorCodeTest.kt
package com.edgeveda.sdk

import org.junit.Assert.*
import org.junit.Test

/**
 * Unit tests for NativeErrorCode — verifies the JNI integer error code mapping to
 * Kotlin EdgeVedaException subclasses. Mirrors the exception hierarchy tests in the
 * Flutter gold standard (edge_veda_test.dart EdgeVedaException group).
 */
class NativeErrorCodeTest {

    // ── fromCode mapping ──────────────────────────────────────────────────────

    @Test
    fun `fromCode maps 0 to OK`() =
        assertEquals(NativeErrorCode.OK, NativeErrorCode.fromCode(0))

    @Test
    fun `fromCode maps -1 to INVALID_PARAM`() =
        assertEquals(NativeErrorCode.INVALID_PARAM, NativeErrorCode.fromCode(-1))

    @Test
    fun `fromCode maps -2 to OUT_OF_MEMORY`() =
        assertEquals(NativeErrorCode.OUT_OF_MEMORY, NativeErrorCode.fromCode(-2))

    @Test
    fun `fromCode maps -3 to MODEL_LOAD_FAILED`() =
        assertEquals(NativeErrorCode.MODEL_LOAD_FAILED, NativeErrorCode.fromCode(-3))

    @Test
    fun `fromCode maps -4 to BACKEND_INIT_FAILED`() =
        assertEquals(NativeErrorCode.BACKEND_INIT_FAILED, NativeErrorCode.fromCode(-4))

    @Test
    fun `fromCode maps -5 to INFERENCE_FAILED`() =
        assertEquals(NativeErrorCode.INFERENCE_FAILED, NativeErrorCode.fromCode(-5))

    @Test
    fun `fromCode maps -6 to CONTEXT_INVALID`() =
        assertEquals(NativeErrorCode.CONTEXT_INVALID, NativeErrorCode.fromCode(-6))

    @Test
    fun `fromCode maps -7 to STREAM_ENDED`() =
        assertEquals(NativeErrorCode.STREAM_ENDED, NativeErrorCode.fromCode(-7))

    @Test
    fun `fromCode maps -8 to NOT_IMPLEMENTED`() =
        assertEquals(NativeErrorCode.NOT_IMPLEMENTED, NativeErrorCode.fromCode(-8))

    @Test
    fun `fromCode maps -9 to MEMORY_LIMIT_EXCEEDED`() =
        assertEquals(NativeErrorCode.MEMORY_LIMIT_EXCEEDED, NativeErrorCode.fromCode(-9))

    @Test
    fun `fromCode maps -10 to UNSUPPORTED_BACKEND`() =
        assertEquals(NativeErrorCode.UNSUPPORTED_BACKEND, NativeErrorCode.fromCode(-10))

    @Test
    fun `fromCode maps -999 to UNKNOWN`() =
        assertEquals(NativeErrorCode.UNKNOWN, NativeErrorCode.fromCode(-999))

    @Test
    fun `fromCode maps unrecognised negative code to UNKNOWN`() =
        assertEquals(NativeErrorCode.UNKNOWN, NativeErrorCode.fromCode(-42))

    @Test
    fun `fromCode maps unrecognised positive code to UNKNOWN`() =
        assertEquals(NativeErrorCode.UNKNOWN, NativeErrorCode.fromCode(999))

    // ── toException: non-error codes return null ───────────────────────────────

    @Test
    fun `OK toException returns null`() =
        assertNull(NativeErrorCode.OK.toException())

    @Test
    fun `STREAM_ENDED toException returns null`() =
        assertNull(NativeErrorCode.STREAM_ENDED.toException())

    // ── toException: error codes return correct exception types ───────────────

    @Test
    fun `INVALID_PARAM toException returns InvalidConfiguration`() {
        val ex = NativeErrorCode.INVALID_PARAM.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.InvalidConfiguration)
    }

    @Test
    fun `OUT_OF_MEMORY toException returns OutOfMemoryError`() {
        val ex = NativeErrorCode.OUT_OF_MEMORY.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.OutOfMemoryError)
    }

    @Test
    fun `MODEL_LOAD_FAILED toException returns ModelLoadError`() {
        val ex = NativeErrorCode.MODEL_LOAD_FAILED.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.ModelLoadError)
    }

    @Test
    fun `BACKEND_INIT_FAILED toException returns ModelLoadError`() {
        val ex = NativeErrorCode.BACKEND_INIT_FAILED.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.ModelLoadError)
    }

    @Test
    fun `INFERENCE_FAILED toException returns GenerationError`() {
        val ex = NativeErrorCode.INFERENCE_FAILED.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.GenerationError)
    }

    @Test
    fun `CONTEXT_INVALID toException returns ContextOverflowError`() {
        val ex = NativeErrorCode.CONTEXT_INVALID.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.ContextOverflowError)
    }

    @Test
    fun `NOT_IMPLEMENTED toException returns UnsupportedOperationError`() {
        val ex = NativeErrorCode.NOT_IMPLEMENTED.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.UnsupportedOperationError)
    }

    @Test
    fun `MEMORY_LIMIT_EXCEEDED toException returns OutOfMemoryError`() {
        val ex = NativeErrorCode.MEMORY_LIMIT_EXCEEDED.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.OutOfMemoryError)
    }

    @Test
    fun `UNSUPPORTED_BACKEND toException returns InvalidConfiguration`() {
        val ex = NativeErrorCode.UNSUPPORTED_BACKEND.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.InvalidConfiguration)
    }

    @Test
    fun `UNKNOWN toException returns NativeError`() {
        val ex = NativeErrorCode.UNKNOWN.toException()
        assertNotNull(ex)
        assertTrue(ex is EdgeVedaException.NativeError)
    }

    @Test
    fun `toException message contains custom context when provided`() {
        val ex = NativeErrorCode.INFERENCE_FAILED.toException("model inference blew up")
        assertNotNull(ex)
        assertTrue(ex!!.message!!.contains("model inference blew up"))
    }

    // ── throwIfError ──────────────────────────────────────────────────────────

    @Test
    fun `throwIfError does not throw for OK`() {
        NativeErrorCode.OK.throwIfError() // must not throw
    }

    @Test
    fun `throwIfError does not throw for STREAM_ENDED`() {
        NativeErrorCode.STREAM_ENDED.throwIfError() // must not throw
    }

    @Test(expected = EdgeVedaException.GenerationError::class)
    fun `throwIfError throws GenerationError for INFERENCE_FAILED`() {
        NativeErrorCode.INFERENCE_FAILED.throwIfError()
    }

    @Test(expected = EdgeVedaException.ModelLoadError::class)
    fun `throwIfError throws ModelLoadError for MODEL_LOAD_FAILED`() {
        NativeErrorCode.MODEL_LOAD_FAILED.throwIfError()
    }

    @Test(expected = EdgeVedaException.OutOfMemoryError::class)
    fun `throwIfError throws OutOfMemoryError for OUT_OF_MEMORY`() {
        NativeErrorCode.OUT_OF_MEMORY.throwIfError()
    }

    // ── enum invariants ───────────────────────────────────────────────────────

    @Test
    fun `OK has code 0`() =
        assertEquals(0, NativeErrorCode.OK.code)

    @Test
    fun `all error codes except OK and STREAM_ENDED have negative code values`() {
        val nonErrors = setOf(NativeErrorCode.OK, NativeErrorCode.STREAM_ENDED)
        NativeErrorCode.entries.filter { it !in nonErrors }.forEach { code ->
            assertTrue("${code.name} should have negative code", code.code < 0)
        }
    }

    @Test
    fun `enum contains exactly 12 entries`() =
        assertEquals(12, NativeErrorCode.entries.size)
}
